### ISSUE TITLE
Table1 sampling time, and evaluation pre-processing

### ISSUE THREAD

**Issue by LuoyaoChen** [cpeng93/DiffuseRecon#3]
Hi!

I am writing regarding two issues I have recently encountered while doing the sampling evaluation. Sampling was using scripts/image_sample_complex_duo.py, the FastMRI dataset:

1. currently, on average, I am encountering about 8 min per slice reconstruction. For each volume in the validation set, even discarding the first 5 slices, we still have 30 slices/volume remaining, considering the eval data contains 200 volumes, it will take too long before I can evaluate the entire 200 volumes. Could you share some details regarding what dataset was used to generate Table 1 (how many **volumes** (i.e. average across the entire 200 in validation set?), and within each volume, was it only the **middle 15 slices** instead of the entire 30 slices?)

2. In addition, it seems like the performance I obtained was somewhat worse than the original Table 1 - Did I miss any procedures while pre-processing the reconstructed image? (I used 8x mask, the pre-trained checkpoint provided in the GitHub during this evaluation).

```
def normalize_complex(data, eps=0.):
    mag = np.abs(data)
    mag_std = mag.std()
    return data / (mag_std + eps), mag_std

## image_true pre-process
image_true = pickle.load(open('../FastMRI/pre_processed_data/val/file1000000_16.pt)', 'rb'))['img'] # here, pre-processed ground truth was generated using the "data_process.py"
image_true, _ = normalize_complex(image_true)
image_true = np.abs(image_true)
image_true = image_true/np.max(image_true)

## image_pred pre-process
out_sample = pickle.load(open(
    '../DiffuseRecon/experiments/img_space_dual/test/file1000000_16.pt)', 'rb'))['fine'] # 4, 320, 320
image_test = out_sample[0, 0]+out_sample[0, 1]*1j
image_test, _ = normalize_complex(image_test)
image_test = np.abs(image_test)
image_test = image_test/np.max(image_test) # I am following line 82 as image_sample_complex_duo.py

## check data range
print(np.max(image_true)-np.min(image_true)) # 0.9996731388254091
print(np.max(image_test)-np.min(image_test)) # 0.9997807

## psnr metric
psnr = skimage.metrics.peak_signal_noise_ratio(
    image_true=image_true, image_test=image_test, data_range=1)
## ssim metric
ssim = skimage.metrics.structural_similarity(image_true, image_test)
print(f"psnr = {psnr}, ssim = {ssim}") 

```

I got 
```
# psnr = 26.231052034619374, ssim = 0.6939833731682563
```
While in the Table1, using a 8x mask, it was around 

```
PSNR ~ 30, SSIM ~0.614
```

Thank you very much for your clarifications!

**Comment by cpeng93** [comment]
Dear Luoyao,

I have uploaded the X8 results from my end in this onedrive link: https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/cpeng26_jh_edu/EabJvNOEeItOjZPhFPWLK98B6qusXEmoDV2YWq0GHNJ_gQ?e=EGrqN5

The volumes are slice-wise normalized, the average metric is calculated based on slice-wise PSNR values. There are several other ways to do this without normalization (e.g. taking the data range of zero-filled reconstruction and apply to our reconstruction), but we didn't find a lot of changes in one way or the other. 

For evaluation, the results above used 100 steps at the coarse stage. In general diffusion model does take a longer time, we used A5000 for our evaluation, so I cannot speak on if your sampling time is correct or not without more information. 

Best,

**Comment by LuoyaoChen** [comment]
Hi!

Thanks for replying and sharing the file - just to make sure, regarding the total number of images reconstruction for evaluation, did you re-construct every single slice per volume? 

Also, how long in total did it take you to re-construct 1 volume, on average?

Thank you so much for sharing! 

Sincerely,

**Comment by cpeng93** [comment]
If you look at the uploaded results, 25 slices are inferred, starting from the second slice down. On average per slice it’s around 20-40 secs. One thing we did was to generate two slices at a time which allows us to improve the efficiency per sampling process.

**Comment by LuoyaoChen** [comment]
Hello, Dr.Peng!

Thank you very much for sharing the reconstruction result, and details regarding the slices used for evaluation! 
### Evaluation Results I obtained: 
I downloaded the reconstructions 8X results you shared via one drive, and evaluated them vs the pre-processed val/ data ground truth. I evaluated your **the shared reconstructed volumes in One Drive**, and obtained,
```
Author provided recon evaluated results: average PSNR = 27.406, average SSIM = 0.527
 ``` 
![image](https://user-images.githubusercontent.com/80808548/214913472-89c77f89-adde-401d-86f7-38f2729f9cd2.png)

which is **lower than in Table1**, 
```
i.e, for 8X, Table1 "PSNR =29.94, SSIM = 0.614
```

That said, I understand it might take a while before the evaluation code can be uploaded; If possible, would you mind checking whether I **missed/did things unnecessary** that led to low results?

### The steps I took for evaluation are: 
1. **Ground truth slices**: pre-processed val dataset downloaded from DiffuseRecon, after that, images were generated using data_process.py
2. Post-processing, **slice-wise normalization**:  
```
recon =np.abs(recon)
recon = (recon - mean(recon))/std(recon)
```
3. Evaluate **only the 2-27 slices**, as per your previous comment "25 slices are inferred, starting from the second slice down"
4. **Packages & data_range**:
for PSNR:
```
psnr = skimage.metrics.peak_signal_noise_ratio(image_true=image_true, image_test=image_test, data_range=7.63)
```
I set **data_range=7.63**, because this is the **average normalized value, across train slices**.
for SSIM
```
ssim = skimage.metrics.structural_similarity(image_true, image_test, data_range=7.63) # 7.21
```
And this is how I obtained the ~27.5 PSNR. Besides above, I have experimented with different PSNR packages (Matlab, and implementations like in [DuroNet](https://github.com/bbbbbbzhou/DuDoRNet/blob/5e47030e373aec066f6cc7957e913e984407b1b2/models/utils.py#L78)), different data_range, and post-processing methods (min-max normalization; divide by max() only; divide by std() only), those performances was either around 27.5, or even lower (using Matlab, for example). 

I would really appreciate it if you could take a look and point out if I missed anything/did anything unnecessary!

Thank you sooo much! Apologies for the long reading!!

Sincerely,
Luoyao

---
### For reference, in cases needed, below is the complete code I used for eval:
```
def calculate_psnr_ssim(pred, gt):
    # image_test, post_processing
    image_test = (pred - np.mean(pred))/np.std(pred)    
    
    # image ground truth: post_processing complex input
    image_true = np.abs(gt)
    image_true = (image_true - np.mean(image_true))/np.std(image_true)
    image_true_range = np.max(image_true) - np.min(image_true)
    
    # psnr
    psnr = skimage.metrics.peak_signal_noise_ratio(image_true=image_true, image_test=image_test, data_range=7.638881333438811) # I also tried 7.21, gave lower performance
    # ssim
    ssim = skimage.metrics.structural_similarity(image_true, image_test, data_range=7.638881333438811) # 7.21
    return psnr, ssim, image_true_range
 
'''
author provided prediction eval, range = 7.63
'''
import glob
all_filenames = sorted([elem[:-3] for elem in os.listdir('/data/data/FastMRI/pre_processed_data/val')])
all_psnr_lis, all_ssim_lis, image_true_range_all_lis = [],[], []
for filename in all_filenames:
    psnr_lis, ssim_lis,image_true_range_lis = [], [], []
    res = glob.glob(
        f'/data/data/FastMRI/pre_processed_data/val/{filename}_*.pt')
    res = sorted(res, key=lambda elem: int(elem.split('/')[-1][:-3].split('_')[-1]))
    
    for i in range(2,27): # len(res)
        try:
            pred = pickle.load(open(os.path.join('/data/data/FastMRI/recon_author_provided', filename +"_full.pt"), 'rb'))[:,:,i] # real number
            pred = np.float32(pred)
            gt = pickle.load(open(f'/data/data/FastMRI/pre_processed_data/val/{filename}_{i}.pt', 'rb'))['img'] ## complex
            psnr, ssim, image_true_range = calculate_psnr_ssim(pred, gt)
            psnr_lis.append(psnr), ssim_lis.append(ssim), image_true_range_lis.append(image_true_range)
        except:
            print(f"filename = {filename}, slice = {i} skipped")
            continue
    print(f"{filename}, len = {len(res)}, avg image_tru_range = {np.average(image_true_range_lis)}, average psnr, ssim = {np.average(psnr_lis), np.average(ssim_lis)}") # 5~33 average = average psnr, ssim for file1000000 = 27.160685033237296, 0.25858155118447745
    all_psnr_lis += psnr_lis
    all_ssim_lis+=ssim_lis
    image_true_range_all_lis += image_true_range_lis
print(f"average psnr, ssim for all = {np.average(all_psnr_lis)}, {np.average(all_ssim_lis)}") # 5~33 average = average psnr, ssim for file1000000 = 27.160685033237296, 0.25858155118447745
print(f"average image_true_range_all_lis = {np.average(image_true_range_all_lis)}")

# plot
plt.figure(figsize=(10,5))
plt.plot(np.arange(len(all_psnr_lis)), all_psnr_lis, label='psnr')
plt.hlines(y=np.average(all_psnr_lis), xmin=0, xmax=len(all_psnr_lis), color='r', label=f'avg {round(np.average(all_psnr_lis),3)}')
plt.xlabel('val data, slice count')
plt.ylabel('PSNR per slice')
plt.title('PSNR per slice, val dataset')
plt.legend()
 
```
I got
```
average psnr, ssim for all = 27.406157770224617, 0.5270564835207837
average image_true_range_all_lis = 7.562108009092217
```

**Comment by cpeng93** [comment]
I believe we used the data range of groundtruth per slice instead of the same data range across all data.

**Comment by LuoyaoChen** [comment]
Hello, Dr.Peng,

I also did experiment with per-slice data range, i.e. 
```

psnr = skimage.metrics.peak_signal_noise_ratio(image_true=image_true, image_test=image_test, data_range = (np.max(image_true)-np.min(image_true)))

```
but obtained similar lower results (PSNR = 27.xx)

I suspect the mismatch (27 vs 29) originated from some tiny discrepancy.  Would you mind sharing the **normalization method** used(z-normalization? min-max, divide by max(), or divide by std()? etc.), and the **packages used** during the evaluation? 

**Also, if convenient, could you please confirm whether any of my steps (1-4) were incorrect/unnecessary?** I have tried several experiments, and have been stuck for a while, since none of the evaluation exceeded 27.5 psnr, 0.527 ssim.

Just in case, you could refer to the attached code above should my explanation is unclear!

Thank you so much! Much appreciated!
Sincerely,

**Comment by cpeng93** [comment]
I've uploaded evaluation code for the test set, try it out on the provided results and see if that resolves your discrepancy.
